main.go

```diff
@@ -98,22 +96,33 @@ func runApp(ctx *cli.Context) error {
 		log.Infof("%s -> %s", topic, worker)
 	}
 
-	Listen(queue, time.Tick(time.Duration(frequency)*time.Millisecond))
+	Listen(queue, time.Duration(ctx.Int64("freq"))*time.Millisecond)
 	return nil
 }
 
 // Listen does the work. It only returns if we get a signal
-func Listen(queue Queue, freq <-chan time.Time) {
+func Listen(queue Queue, freq time.Duration) {
+	messages := make(chan *Message, 10)
+
+	go pollForever(queue, messages, freq)
+
 	for {
 		select {
 		case <-signals:
 			log.Info("Got TERM")
 			queue.Semaphore().Wait()
 			return
-		case tick := <-freq:
-			log.Debug("Polling at: ", tick)
+		case msg := <-messages:
 			queue.Semaphore().Add(1)
-			go queue.Poll()
+			go queue.ProcessMessage(msg)
 		}
 	}
 }
+
+func pollForever(queue Queue, messages chan *Message, freq time.Duration) {
+	for {
+		throttle := time.After(freq)
+		queue.Poll(messages)
+		<-throttle
```

Comment:
The concurrency based testing I did for `Listen()` is already dark magic. Testing this would require me trying to ensure that something is going to finish in under a certain frequency.